Add E2E test suite

orelmisan opened the pull request:
Add a ginkgo[1] based E2E test suite its dependencies.

Test checkup Job configuration and successful execution.
The E2E test is executed by:

```bash
make e2e-test
```

Or:

```bash
cd tests
ginkgo
```

The test suite respects the following environment variables:
`TEST_NAMESPACE` - set the namespace on which to perform the tests
`TEST_IMAGE` - sets the image name and tag, on which to perform the tests

The `e2e-test` target assumes that the kubeconfig file is under `~/.kube/`.

Tested against an OpenShift 4.11 cluster.

[1] https://onsi.github.io/ginkgo/

orelmisan commented:
The linter is passing locally, and is failing in the CI because of timeout.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/06452a3e0ea1b9a9f3b0595523403d4131c41aae..fa124fa4feede52b90b5b033ad854e5bc234ff41): Limit the linter to `cmd` and `tests` directories.
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/fa124fa4feede52b90b5b033ad854e5bc234ff41..aef674f830cab1d66da7f260ae773419ebf3284c): Add a timeout to the linter.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,111 @@
+			}
+
+			return false
+		}, 5*time.Minute, 5*time.Second).Should(BeTrue())
```
In case of failure it will give no valuable informative.
My suggestion is to return the conditions slice and lookup the condition in it. (try to use `ContainElement` and other matchers)
This way, if there is a failure, you will get the list of existing conditions for debugging and a much better message.

orelmisan replied:
Thank you.
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,111 @@
+
+func boolPointer(b bool) *bool {
+	return &b
+}
```
This will be cleaner:
```
func pointer[T any](v T) *T {
	return &v
}
```

orelmisan replied:
Thank you.
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,111 @@
+								SeccompProfile: &corev1.SeccompProfile{
+									Type: corev1.SeccompProfileTypeRuntimeDefault,
+								},
+							},
```
Perhaps extract this to a function to reduce the noise of unrelated data.

orelmisan replied:
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,111 @@
+ *
+ */
+
+package tests_test
```
Why not just `test`?

orelmisan replied:
It was auto-generated by ginkgo.
Fixed to `tests`.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,111 @@
+		})
+	})
+
+	It("should succeed", func() {
```
What does "succeed" mean? The test itself seems to check that it completes.

orelmisan replied:
Changed to `should complete successfully`.

EdDev commented on tests/tests_suite_test.go:
```diff
@@ -0,0 +1,76 @@
+	defaultImageName = "quay.io/kiagnose/kubevirt-rt-checkup:main"
+)
+
+var (
+	client        *kubernetes.Clientset
+	testNamespace string
+	testImageName string
+)
```
Seems unsafe to use globals that anyone can override.

orelmisan replied:
I followed the advice of the ginkgo documentation:
https://onsi.github.io/ginkgo/#suite-setup-and-cleanup-beforesuite-and-aftersuite

Is there another approach you are familiar with?

EdDev reviewed (changes requested):
For such a small project, I do not understand why there is a need to use ginkgo. Even the ref checkups do not use it.
Moreover, I do not understand why e2e/functional tests are introduced before any production & unit tests.

But this is not for me to decide.
I commented on the technical aspects per what I saw.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/aef674f830cab1d66da7f260ae773419ebf3284c..70bfee7a10e17853c63beef172d9a468f6759c59):

Answered @EdDev's [comments](https://github.com/kiagnose/kubevirt-rt-checkup/pull/6#pullrequestreview-1235758195).
Aligned K8s dependencies to v0.25.5

RamLavi commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,114 @@
+					RestartPolicy:      corev1.RestartPolicyNever,
+					Containers: []corev1.Container{
+						{
+							Name:            "rt-checkup",
```
testCheckupJobName?

orelmisan replied:
This is the name of the checkup container, it could be constant (it is constant in the VM latency checkup as well).

RamLavi commented on tests/tests_suite_test.go:
```diff
@@ -17,16 +17,60 @@
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	Expect(err).NotTo(HaveOccurred())
+
+	client, err = kubernetes.NewForConfig(config)
```
in the future you will probably need to kubectl vms, why not use kubevirt's client?

orelmisan replied:
At this stage it is not necessary, so I prefer to postpone using it until I'll have to.

RamLavi reviewed (commented):
Could you add a makefile target for e2e tests? 
suggestions:
```
make functest
make test/e2e
```

orelmisan commented:
> Could you add a makefile target for e2e tests? suggestions:
> 
> ```
> make functest
> make test/e2e
> ```

In the PR description I've mentioned that the E2E tests are currently separated from the Makefile.
I don't mind adding a target, but at the moment there won't be a github actions workflow for it.

orelmisan commented:
Change: Add `test-e2e` target to Makefile.

RamLavi reviewed (dismissed):
/lgtm

orelmisan committed:
Add E2E test suite

Add the ginkgo[1] based E2E test suite and vendor its dependencies.

The `tests_suite_test.go` was generated by:
```bash
ginkgo bootstrap
```

The E2E tests could be executed by:
```bash
cd tests
go test
```

Or:

```bash
cd tests
ginkgo
```

The ginkgo cli tool could be installed using:
```bash
go install github.com/onsi/ginkgo/v2/ginkgo
```

[1] https://onsi.github.io/ginkgo/

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
Makefile, lint: lint the tests directory

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
Makefile: Add e2e-test target

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
E2E test suite: Configure and run the checkup

Create the following objects:
- ServiceAccount
- Role
- RoleBinding
- ConfigMap - User-supplied
- Job - points to the ConfigMap

Create the checkup Job on the remote cluster.
Wait for it to successfully complete.
Check the `status.succeeded` and `status.failureReason` to
make sure the checkup had succeeded.

In order to set the image name and tag:
```bash
TEST_IMAGE=quay.io/myuser/kubevirt-rt-checkup:devel make e2e-test
```

In order to set the namespace:
```bash
TEST_NAMESPACE=my-namespace make e2e-test
```

The `e2e-test` target assumes the kubeconfig file is under
`$HOME/.kube`.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/08114eaaaa0c2612b32c3ce0ae13217f3af6e81c..e3784fa7667aeee59e537fbe21f272fa7d441d9e):
1. Rebase
2. Create ServiceAccount, Role, RoleBinding and ConfigMap.
3. Point the checkup to the ConfigMap.
4. Test the ConfigMap data at checkup completion.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/e3784fa7667aeee59e537fbe21f272fa7d441d9e..33c15cd33647b8415c7ad059e43196d6866b4f2f): Fixed a white space.

RamLavi reviewed (dismissed):
/lgtm

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,237 @@
+)
+
+var _ = Describe("Execute the checkup Job", func() {
+	var checkupServiceAccount *corev1.ServiceAccount
+	var kiagnoseConfigMapAccessRole *rbacv1.Role
+	var kiagnoseConfigMapAccessRoleBinding *rbacv1.RoleBinding
+	var configMap *corev1.ConfigMap
+	var checkupJob *batchv1.Job
```
nit:
```
var (
...
)
```

orelmisan replied:
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,237 @@
+		_, err = client.BatchV1().Jobs(testNamespace).Create(context.Background(), checkupJob, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		DeferCleanup(func() {
```
I do not think this block is correct.
If a creation of an object was successful, then you need to use a deferred cleanup.
Not group them all at the end.

As an example, if the job create failed, no one will cleanup the configmap and RBAC items.

orelmisan replied:
Added DeferCleanup() after each successful object creation.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,237 @@
+	testCheckupJobName                         = "rt-checkup"
+)
+
+var _ = Describe("Execute the checkup Job", func() {
```
It reads like a verb/action, but it needs to be a description of a "thing".
E.g. "checkup execution".

orelmisan replied:
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,237 @@
+		})
+	})
+
+	It("should complete successfully", func() {
```
It is sad that I need to go 60 lines above to understand what should complete successfully.

orelmisan replied:
Moved the permission setup to a dedicated function.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,237 @@
+		configMap, err := client.CoreV1().ConfigMaps(testNamespace).Get(context.Background(), testConfigMapName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(configMap.Data).NotTo(BeNil())
+		Expect(configMap.Data["status.succeeded"]).To(Equal("true"))
+		Expect(configMap.Data["status.failureReason"]).To(BeEmpty())
```
It will be hard to debug a failure here.
We will not see the whole picture of what this config-map contains.

If it is not possible to assert the whole object, then at least report its whole content on failure (through the assert description possibly).

orelmisan replied:
> then at least report its whole content on failure (through the assert description possibly).

Could you please explain how this should be done?

EdDev replied:
The matchers have a "description" option.
Example: `Expect(prepare.executed).To(BeTrue(), "the prepare step should execute")`

In that text you can add more info, like the content of all the configmap.

orelmisan replied:
This line prints the CM data on failure:
```go
Expect(configMap.Data["status.succeeded"]).To(Equal("true"), fmt.Sprint(configMap.Data))
```
WDYT?

EdDev replied:
Yes, something like that. Maybe add “should succeed” for some context. Up to you.

orelmisan replied:
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/33c15cd33647b8415c7ad059e43196d6866b4f2f..a1797706e42a08010d0d29508854745a69c729e0): 
1. Move the permission setup to a dedicated function.
2. Add a DeferCleanup() after each object's successful creation.
3.  Replaced two string literals with constants from the `rbacv1` package.

EdDev commented on tests/checkup_test.go:
```diff
@@ -0,0 +1,254 @@
+	_, err = client.CoreV1().ServiceAccounts(testNamespace).Create(context.Background(), checkupServiceAccount, metav1.CreateOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	DeferCleanup(func() {
```
Interesting.. so this defer can work even if it is in a helper function?

orelmisan replied:
It cleans up the RoleBinding, Role and ServiceAccount.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/a1797706e42a08010d0d29508854745a69c729e0..85652618e2c28282ef63037f76de272483e315e7): Answered @EdDev's [comment](https://github.com/kiagnose/kubevirt-rt-checkup/pull/6#discussion_r1068083440).

EdDev reviewed (approved):
Thanks
